### ISSUE TITLE
Using naming convention token `referred_column_0_name` raises `sqlalchemy.exc.NoReferencedColumnError`

### ISSUE THREAD

**Issue by leonarduschen** [sqlalchemy/sqlalchemy#7958]
### Describe the bug

Using naming convention token `referred_column_0_name` raises `sqlalchemy.exc.NoReferencedColumnError`

### To Reproduce

```python
from sqlalchemy import Column, ForeignKey, Integer, MetaData
from sqlalchemy.orm import declarative_base, relationship

metadata = MetaData(
    naming_convention={
        "fk": "fk_%(referred_column_0_name)s",
    }
)

Base = declarative_base(metadata=metadata)


class Person(Base):
    __tablename__ = "persons"

    id = Column(Integer, primary_key=True)

    addresses = relationship("addresses", back_populates="person")


class Addresses(Base):
    __tablename__ = "addresses"

    id = Column(Integer, primary_key=True, nullable=False)
    person_id = Column(ForeignKey("persons.id"))

    person = relationship("persons", back_populates="addresses")
```

### Error

```
# Copy complete stack trace and error message here, including SQL log output if applicable.
``` 

```python
Traceback (most recent call last):
  File "/home/leonarduschen/src/test/main.py", line 21, in <module>
    class Addresses(Base):
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/orm/decl_api.py", line 72, in __init__
    _as_declarative(reg, cls, dict_)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 126, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 177, in setup_mapping
    return cfg_cls(registry, cls_, dict_, table, mapper_kw)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 322, in __init__
    self._setup_table(table)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 811, in _setup_table
    table_cls(
  File "<string>", line 2, in __new__
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 309, in warned
    return fn(*args, **kwargs)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 616, in __new__
    metadata._remove_table(name, schema)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 611, in __new__
    table._init(name, metadata, *args, **kw)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 697, in _init
    self._init_items(
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 144, in _init_items
    spwd(self, **kw)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 1047, in _set_parent_with_dispatch
    self.dispatch.after_parent_attach(self, parent)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/event/attr.py", line 343, in __call__
    fn(*args, **kw)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 2576, in _set_table
    self.constraint._set_parent_with_dispatch(table)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 1047, in _set_parent_with_dispatch
    self.dispatch.after_parent_attach(self, parent)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/event/attr.py", line 256, in __call__
    fn(*args, **kw)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/naming.py", line 208, in _constraint_name
    newname = _constraint_name_for_table(const, table)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/naming.py", line 166, in _constraint_name_for_table
    convention
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/naming.py", line 126, in __getitem__
    return getattr(self, attr)(idx)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/naming.py", line 95, in _key_referred_column_X_name
    return fk.column.name
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 1113, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/home/leonarduschen/src/test/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 2521, in column
    raise exc.NoReferencedColumnError(
sqlalchemy.exc.NoReferencedColumnError: Could not initialize target column for ForeignKey 'persons.id' on table 'addresses': table 'persons' has no column named 'id'
```

### Versions

- OS: Pop!_OS 21.10 x86_64
- Python: 3.9.7
- SQLAlchemy: 1.4.35

**Comment by zzzeek** [comment]
this bug exists to the introduction of this naming convention feature, and the feature is currently only tested with ``ForeignKeyConstraint``, not ``ForeignKey``.  So you are likely the first person to actually use this particular naming convention feature in any real-world use case.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**backport 6f02d5edd88fe247 to 1.4** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3810

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**forwards port test cases and changelog for #7958 to 2.0** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3811

**Comment by leonarduschen** [comment]
Thanks for the fix @zzzeek! I'm guessing it'll be released on `1.4.36`?

**Comment by zzzeek** [comment]
yes it's all ready for when i put 1.4.36 out.   the feature has never worked in a practical sense since it crashes with ForeignKey.
